### PR TITLE
Gke support

### PR DESCRIPTION
**Pull request by haiodo**
Google Cloud Engine changes.

## Description
Google Cloud Engine changes.

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [x] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by haiodo** [comment]
@edwarnicke I'm able to deploy most of NSC/NSE to GKE Kubernetes.
After rebasing to latest master, with some cpu limit fixes and NSC yaml fix, some of NSC looks OK, but could not ping, so will continue checking tomorrow.

**Comment by nickolaev** [comment]
Shall we close #798?

File: .mk/k8s.mk [hunk 1]
```diff
@@ -50,9 +59,25 @@ include .mk/kind.mk
 # before running make
 include .mk/null.mk
 
+include .mk/docker.mk
+include .mk/gcb.mk
+
```
Comment (nickolaev) [inline .mk/k8s.mk]: Should these be included conditionaly?

File: .mk/docker.mk [hunk 2]
```diff
@@ -12,10 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
 BUILD_CONTAINERS=nsmd nsmdp nsmd-k8s vppagent-dataplane vppagent-dataplane-dev
 BUILD_CONTAINERS+=devenv crossconnect-monitor
 BUILD_CONTAINERS+=nsc icmp-responder-nse
 BUILD_CONTAINERS+=vppagent-firewall-nse
+
+
```
Comment (nickolaev) [inline .mk/docker.mk]: Empty lines?

File: .mk/k8s.mk [hunk 3]
```diff
@@ -12,6 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+BUILD_CONTAINERS=nsmd nsmdp nsmd-k8s vppagent-dataplane
+BUILD_CONTAINERS+=devenv crossconnect-monitor
+BUILD_CONTAINERS+=nsc icmp-responder-nse vppagent-icmp-responder-nse vppagent-nsc
+BUILD_CONTAINERS+=vppagent-firewall-nse
+BUILD_CONTAINERS+=admission-webhook
+
```
Comment (nickolaev) [inline .mk/k8s.mk]: These are already in `.mk/docker.mk`.

File: .mk/k8s.mk [hunk 4]
```diff
@@ -75,25 +100,30 @@ k8s-deployonly: $(addsuffix -deployonly,$(addprefix k8s-,$(DEPLOYS)))
...
-	@sed "s;\(image:[ \t]*networkservicemesh/[^:]*\).*;\1$${COMMIT/$${COMMIT}/:$${COMMIT}};" ${K8S_CONF_DIR}/admission-webhook.yaml | kubectl apply -f -
+	@sed "s;\(image:[ \t]*\)\(networkservicemesh\)\(/[^:]*\).*;\1${CONTAINER_REPO}\3$${COMMIT/$${COMMIT}/:$${COMMIT}};" ${K8S_CONF_DIR}/admission-webhook.yaml | kubectl apply -f -
 	@echo "Installing webhook..."
 	@cat ./k8s/conf/admission-webhook-cfg.yaml | ./scripts/webhook-patch-ca-bundle.sh | kubectl apply -f -
 
+.PHONY: k8s-vpn-gateway-nse-deploy
+k8s-vpn-gateway-nse-deploy: k8s-start k8s-config k8s-%-delete k8s-%-load-images
+	@until ! $$(kubectl get pods | grep -q ^vpn-gateway-nse ); do echo "Wait for vpn-gateway-nse to terminate"; sleep 1; done
+	@sed "s;\(image:[ \t]*\)\(networkservicemesh\)\(/icmp-responder-nse[^:]*\).*;\1${CONTAINER_REPO}\3$${COMMIT/$${COMMIT}/:$${COMMIT}};" ${K8S_CONF_DIR}/vpn-gateway-nse.yaml | kubectl apply -f -
+
```
Comment (nickolaev) [inline .mk/k8s.mk]: Is this a GKE specific piece? If not, please split in separate PR.

File: .mk/k8s.mk [hunk 5]
```diff
@@ -148,7 +178,10 @@ k8s-build: $(addsuffix -build,$(addprefix k8s-,$(DEPLOYS)))
 k8s-jaeger-build:
 
 .PHONY: k8s-jaeger-save
-k8s-jaeger-save:
+k8s-jaeger-build:
```
Comment (nickolaev) [inline .mk/k8s.mk]: Update the `.PHONY` target before this line?

File: .mk/k8s.mk [hunk 6]
```diff
@@ -188,9 +221,7 @@ k8s-secure-intranet-connectivity-build:
 k8s-secure-intranet-connectivity-save:
 
 .PHONY: k8s-secure-intranet-connectivity-load-images
-k8s-secure-intranet-connectivity-load-images:
-	@echo "Wait for nsmd to register the resources"
-	@sleep 10
+k8s-secure-intranet-connectivity-load-images: ;
```
Comment (nickolaev) [inline .mk/k8s.mk]: That sleep is unnecessary indeed. Do we need the semicolon in the end?

File: docker/Dockerfile.devenv [hunk 7]
```diff
@@ -0,0 +1 @@
+./debug/Dockerfile.debug
```
Comment (nickolaev) [inline docker/Dockerfile.devenv]: Not GKE related, please remove.

File: docs/gke-quickstart.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+* [Quick start guides](https://cloud.google.com/sdk/docs/quickstarts)
+
+### Macos via brew cask
+`brew cask install google-cloud-sdk`
+
+## Use Google Cloud environment.
+
+Before start using of GKE please execute following command:
+
+`source ./.env/gke.env`
```
Comment (nickolaev) [inline docs/gke-quickstart.md]: Can we have these in pre-formated blocks:
\```shell
 source ./.env/gke.env
\```

File: gcb/cloudbuild.yaml [hunk 9]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (nickolaev) [inline gcb/cloudbuild.yaml]: Please move these under `deployments/gcb/`.

File: k8s/conf/nsc.yaml [hunk 10]
```diff
@@ -8,8 +8,6 @@ spec:
       labels:
         networkservicemesh.io/app: "icmp-responder-nsc"
     spec:
-      nodeSelector:
```
Comment (nickolaev) [inline k8s/conf/nsc.yaml]: Hmm is this safe? Have you tried running this locally?

File: k8s/conf/cpu-defaults.yaml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
```
Comment (nickolaev) [inline k8s/conf/cpu-defaults.yaml]: Is this GKE specific?

File: .mk/k8s.mk (hunk 4 above)
Comment (haiodo) [reply .mk/k8s.mk]: @edwarnicke did you remember why you was need this part?

File: k8s/conf/nsc.yaml (hunk 10 above)
Comment (haiodo) [reply k8s/conf/nsc.yaml]: It is safe I suppose, since we should not depend for demo container to belong to any node, and inside GKE there is no master node at all, all workers are similar. I will try to find some analogies.

File: k8s/conf/cpu-defaults.yaml (hunk 11 above)
Comment (haiodo) [reply k8s/conf/cpu-defaults.yaml]: GKE is more strict into defaults.

**Comment by haiodo** [comment]
Hi @edwarnicke  today fight with #849, not yet able to make it working :-(

File: .mk/k8s.mk (hunk 4 above)
Comment (edwarnicke) [reply .mk/k8s.mk]: @nickolaev Here's the deal... you need to get the containers you've built to the K8s cluster somehow... the most expeditious is to build them in Google Cloud Builder and get them from a repo there.  This just provides the extra knobs to enable that.  I suspect we'd do this for every public cloud... but this mechanism should work for all of them.

File: k8s/conf/nsc.yaml (hunk 10 above)
Comment (edwarnicke) [reply k8s/conf/nsc.yaml]: @haiodo Yeah... pegging to nodes by name is fragile... it's part of why I used to do this with Pod affinities and anti-affinities.

File: k8s/conf/cpu-defaults.yaml (hunk 11 above)
Comment (haiodo) [reply k8s/conf/cpu-defaults.yaml]: It is a good defaults, enabled only for GKE, possible some other also need them.

File: .mk/k8s.mk (hunk 4 above)
Comment (haiodo) [reply .mk/k8s.mk]: This is generic, yes, I've also updated integration tests to use same environment variable to fix pods deployed.
